performance/vip-tweaks.php

```diff
@@ -25,50 +39,118 @@ function wpcom_vip_disable_performance_tweaks() {
-		// the new attachment is not in the same month/year as the data in our transient
-		delete_transient( 'wpcom_media_months_array' );
+/**
+ * Query wrapper that caches the results of available months given a particular post type.
+ *
+ * See WP_List_Table::months_dropdown() & wp_enqueue_media() in WP core.
+ *
+ * @param string $post_type
+ * @return object[]
+ */
+function wpcom_vip_get_available_months_for_filters( $post_type ) {
+	global $wpdb;
+
+	$cache_key = "available_filter_months_$post_type";
+	$months    = wp_cache_get( $cache_key, 'vip' );
+	if ( is_array( $months ) ) {
+		// Happiest-path, cache exists already :).
+		return $months;
 	}
+
+	$extra_checks = '';
+	if ( 'attachment' !== $post_type ) {
+		$extra_checks = " AND post_status != 'auto-draft' AND post_status != 'trash'";
+	}
+
+	// phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+	$months = $wpdb->get_results(
+		$wpdb->prepare(
+			"
+			SELECT DISTINCT YEAR( post_date ) AS year, MONTH( post_date ) AS month
+			FROM $wpdb->posts
+			WHERE post_type = %s
+			$extra_checks
+			ORDER BY post_date DESC
+			",
+			$post_type
+		)
+	);
+	// phpcs:enable WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+
+	wp_cache_set( $cache_key, $months, 'vip', DAY_IN_SECONDS );
+	return $months;
 }
 
-if ( is_admin() ) {
-	add_filter( 'media_library_show_video_playlist', '__return_true' );
-	add_filter( 'media_library_show_audio_playlist', '__return_true' );
-
-	add_filter( 'media_library_months_with_files', 'wpcom_vip_media_library_months_with_files' );
-	function wpcom_vip_media_library_months_with_files() {
-		global $wpdb;
-
-		$months = get_transient( 'wpcom_media_months_array' );
-
-		if ( false === $months ) {
-			// phpcs:ignore WordPress.DB.DirectDatabaseQuery
-			$months = $wpdb->get_results( $wpdb->prepare( "
-				SELECT DISTINCT YEAR( post_date ) AS year, MONTH( post_date ) AS month
-					FROM $wpdb->posts
-					WHERE post_type = %s
-					ORDER BY post_date DESC
-				", 'attachment' ) );
-			set_transient( 'wpcom_media_months_array', $months );
+/**
+ * Helper method to bust the "available months" cache when needed.
+ *
+ * Note that we only bust the cache when a new month is added,
+ * For deletions, we just let the 24-hour TTL handle that.
+ *
+ * @param string $post_type
+ * @return object[]
+ */
+function wpcom_vip_maybe_bust_available_months_cache( $post_id ) {
+	$post_type = get_post_type( $post_id );
+	$cache_key = "available_filter_months_$post_type";
+
+	$existing_months = wp_cache_get( $cache_key, 'vip' );
+	if ( ! is_array( $existing_months ) ) {
+		// No cache to bust.
+		return false;
+	}
+
+	$post_year  = get_the_time( 'Y', $post_id );
+	$post_month = get_the_time( 'n', $post_id );
+
+	$found = false;
+	foreach ( $existing_months as $date ) {
+		if ( $post_year === $date->year && $post_month === $date->month ) {
+			$found = true;
```

Comment:
I think there should be a `break;` here. I see no reason for the the loop to continue after a match is found. Do you?

Reply:
Nice, good find.